File: sdk/typescript/test/graph.test.ts
import { describe, it, expect } from "vitest";
import { Client, APIError } from "../src/index";
import { mockFetch } from "./helpers";

describe("GraphClient", () => {
  it("Verify sends GET /graph/verify", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/graph/verify",
      respBody: { issues: [] },
    });
    try {
      const c = new Client("http://localhost:8420");
      const got = await c.graph.verify();
      expect(got.issues).toEqual([]);
    } finally {
      restore();
    }
  });

  it("Contradictions sends GET /contradictions (no entity id)", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/contradictions",
      respBody: [{ source_id: "a", source_content: "x", target_id: "b", target_content: "y" }],
    });
    try {
      const c = new Client("http://localhost:8420");
      const got = await c.graph.contradictions("");
      expect(got).toHaveLength(1);
      expect(got[0].source_id).toBe("a");
    } finally {
      restore();
    }
  });

  it("Contradictions appends ?id=<encoded> when entityId is given", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/contradictions",
      query: { id: "paris" },
      respBody: [],
    });
    try {
      const c = new Client("http://localhost:8420");
      await c.graph.contradictions("paris");
    } finally {
      restore();
    }
  });

  it("ConnectedComponents sends GET /connected-components (no min_size)", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/connected-components",
      respBody: [{ ids: ["a", "b", "c"], size: 3, avg_degree: 1.5 }],
    });
    try {
      const c = new Client("http://localhost:8420");
      const got = await c.graph.connectedComponents(0);
      expect(got).toHaveLength(1);
      expect(got[0].size).toBe(3);
    } finally {
      restore();
    }
  });

  it("ConnectedComponents appends ?min_size=N when minSize > 0", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/connected-components",
      query: { min_size: "2" },
      respBody: [],
    });
    try {
      const c = new Client("http://localhost:8420");
      await c.graph.connectedComponents(2);
    } finally {
      restore();
    }
  });

  it("Communities sends GET /communities", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/communities",
      respBody: { communities: [], global_modularity: 0.0, total_communities: 0 },
    });
    try {
      const c = new Client("http://localhost:8420");
      const got = await c.graph.communities({});
      expect(got.total_communities).toBe(0);
    } finally {
      restore();
    }
  });

  it("Timeline sends GET /timeline (no limit)", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/timeline",
      respBody: [{ id: "t1", category: "world", content: "x", created_at: "2024-01-01T00:00:00Z" }],
    });
    try {
      const c = new Client("http://localhost:8420");
      const got = await c.graph.timeline(0);
      expect(got).toHaveLength(1);
    } finally {
      restore();
    }
  });

  it("Timeline appends ?limit=N when limit > 0", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/timeline",
      query: { limit: "10" },
      respBody: [],
    });
    try {
      const c = new Client("http://localhost:8420");
      await c.graph.timeline(10);
    } finally {
      restore();
    }
  });

  it("Provenance sends GET /provenance", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/provenance",
      respBody: [{ id: "e1", category: "world", content: "x", archived: false }],
    });
    try {
      const c = new Client("http://localhost:8420");
      const got = await c.graph.provenance({});
      expect(got).toHaveLength(1);
    } finally {
      restore();
    }
  });

  it("RecoveryPlan appends ?id=<taskId>", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/recovery-plan",
      query: { id: "task-1" },
      respBody: [{ id: "rb-1", category: "task", content: "recover", archived: false }],
    });
    try {
      const c = new Client("http://localhost:8420");
      const got = await c.graph.recoveryPlan("task-1");
      expect(got).toHaveLength(1);
      expect(got[0].id).toBe("rb-1");
    } finally {
      restore();
    }
  });

  it("Error response propagates as APIError", async () => {
    const restore = mockFetch({
      method: "GET",
      path: "/graph/verify",
      status: 500,
      respBody: { error: "db down" },
    });
    try {
      const c = new Client("http://localhost:8420");
      await expect(c.graph.verify()).rejects.toMatchObject({
        statusCode: 500,
        message: expect.stringContaining("db down"),
      });
    } finally {
      restore();
    }
  });
});

File: sdk/go/graph_test.go
package hermem

import (
	"context"
	"net/http"
	"testing"
)

func TestGraphVerify(t *testing.T) {
	runSDKCall(t, "GET", "/graph/verify", http.StatusOK, VerifyReport{Issues: nil}, func(c *Client) {
		got, err := c.Graph.Verify(context.Background())
		if err != nil {
			t.Fatalf("Verify: %v", err)
		}
		if got == nil {
			t.Fatal("got nil")
		}
	})
}

func TestGraphContradictions(t *testing.T) {
	runSDKCall(t, "GET", "/contradictions", http.StatusOK, []ContradictionPair{
		{SourceID: "a", SourceContent: "x", TargetID: "b", TargetContent: "y"},
	}, func(c *Client) {
		got, err := c.Graph.Contradictions(context.Background(), "")
		if err != nil {
			t.Fatalf("Contradictions: %v", err)
		}
		if len(got) != 1 {
			t.Fatalf("got %d, want 1", len(got))
		}
		if got[0].SourceID != "a" {
			t.Errorf("source_id: got %q, want a", got[0].SourceID)
		}
	})
}

// TestGraphContradictionsWithID verifies the SDK sends ?id=<encoded>
// when an entity id is supplied.
func TestGraphContradictionsWithID(t *testing.T) {
	runSDKCallWithQuery(t, "GET", "/contradictions", "id=paris", http.StatusOK, []ContradictionPair{}, func(c *Client) {
		got, err := c.Graph.Contradictions(context.Background(), "paris")
		if err != nil {
			t.Fatalf("Contradictions: %v", err)
		}
		if got == nil {
			t.Fatal("got nil")
		}
	})
}

func TestGraphConnectedComponents(t *testing.T) {
	runSDKCall(t, "GET", "/connected-components", http.StatusOK, []ConnectedComponent{
		{Size: 3, AvgDegree: 1.5},
	}, func(c *Client) {
		got, err := c.Graph.ConnectedComponents(context.Background(), 0) // 0 = no min_size query
		if err != nil {
			t.Fatalf("ConnectedComponents: %v", err)
		}
		if len(got) != 1 || got[0].Size != 3 {
			t.Errorf("got %+v, want one component of size 3", got)
		}
	})
}

// TestGraphConnectedComponentsWithMinSize verifies the SDK sends
// ?min_size=N when minSize > 0.
func TestGraphConnectedComponentsWithMinSize(t *testing.T) {
	runSDKCallWithQuery(t, "GET", "/connected-components", "min_size=2", http.StatusOK, []ConnectedComponent{}, func(c *Client) {
		_, err := c.Graph.ConnectedComponents(context.Background(), 2)
		if err != nil {
			t.Fatalf("ConnectedComponents: %v", err)
		}
	})
}

func TestGraphCommunities(t *testing.T) {
	runSDKCall(t, "GET", "/communities", http.StatusOK, map[string]interface{}{
		"communities":       []interface{}{},
		"global_modularity": 0.0,
		"total_communities": 0,
	}, func(c *Client) {
		got, err := c.Graph.Communities(context.Background(), 0, 0)
		if err != nil {
			t.Fatalf("Communities: %v", err)
		}
		if got == nil {
			t.Fatal("got nil")
		}
	})
}

func TestGraphTimeline(t *testing.T) {
	runSDKCall(t, "GET", "/timeline", http.StatusOK, []TimelineEntry{
		{ID: "t1", Category: "world", Content: "x", CreatedAt: "2024-01-01T00:00:00Z"},
	}, func(c *Client) {
		got, err := c.Graph.Timeline(context.Background(), 0) // 0 = no limit query
		if err != nil {
			t.Fatalf("Timeline: %v", err)
		}
		if len(got) != 1 {
			t.Fatalf("got %d, want 1", len(got))
		}
	})
}

// TestGraphTimelineWithLimit verifies the SDK sends ?limit=N.
func TestGraphTimelineWithLimit(t *testing.T) {
	runSDKCallWithQuery(t, "GET", "/timeline", "limit=10", http.StatusOK, []TimelineEntry{}, func(c *Client) {
		_, err := c.Graph.Timeline(context.Background(), 10)
		if err != nil {
			t.Fatalf("Timeline: %v", err)
		}
	})
}

func TestGraphProvenance(t *testing.T) {
	runSDKCall(t, "GET", "/provenance", http.StatusOK, []Entity{
		{ID: "e1", Category: "world", Content: "x"},
	}, func(c *Client) {
		got, err := c.Graph.Provenance(context.Background(), "", "", "", 0)
		if err != nil {
			t.Fatalf("Provenance: %v", err)
		}
		if len(got) != 1 {
			t.Fatalf("got %d, want 1", len(got))
		}
	})
}

func TestGraphRecoveryPlan(t *testing.T) {
	runSDKCallWithQuery(t, "GET", "/recovery-plan", "id=task-1", http.StatusOK, []Entity{
		{ID: "rb-1", Category: "task", Content: "recover"},
	}, func(c *Client) {
		got, err := c.Graph.RecoveryPlan(context.Background(), "task-1")
		if err != nil {
			t.Fatalf("RecoveryPlan: %v", err)
		}
		if len(got) != 1 {
			t.Fatalf("got %d, want 1", len(got))
		}
	})
}

func TestGraphError(t *testing.T) {
	runSDKCall(t, "GET", "/graph/verify", http.StatusInternalServerError, `{"error":"db down"}`, func(c *Client) {
		_, err := c.Graph.Verify(context.Background())
		if err == nil {
			t.Fatal("expected error, got nil")
		}
		apiErr, ok := err.(*APIError)
		if !ok {
			t.Fatalf("got %T, want *APIError", err)
		}
		if apiErr.StatusCode != 500 {
			t.Errorf("status: got %d, want 500", apiErr.StatusCode)
		}
	})
}
